micrometer-core/src/main/java/io/micrometer/core/instrument/HighCardinalityTagsDetector.java

```diff
@@ -0,0 +1,156 @@
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+import io.micrometer.common.util.internal.logging.InternalLogger;
+import io.micrometer.common.util.internal.logging.InternalLoggerFactory;
+
+/**
+ * Tries to detect high cardinality tags by checking if the amount of Meters with the same
+ * name is above a threshold. You can use this class in two ways: 1. Call findFirst and
+ * check if you get any results, if so you probably have high cardinality tags 2. Call
+ * start which will start a scheduled job that will do this check for you
+ *
+ * @author Jonatan Ivanov
+ * @since 1.10.0
+ */
+public class HighCardinalityTagsDetector implements AutoCloseable {
+
+    private static final InternalLogger LOGGER = InternalLoggerFactory.getInstance(HighCardinalityTagsDetector.class);
+
+    private static final int DEFAULT_THRESHOLD = 1_000_000;
+
+    private static final Duration DEFAULT_DELAY = Duration.ofMinutes(5);
+
+    public static final Consumer<String> DEFAULT_CONSUMER = HighCardinalityTagsDetector::logWarning;
+
+    private final MeterRegistry registry;
+
+    private final int threshold;
+
+    private final Consumer<String> meterNameConsumer;
+
+    private final ScheduledExecutorService scheduledExecutorService;
+
+    private final Duration delay;
+
+    /**
+     * @param registry The registry to use to check the Meters in it
+     */
+    public HighCardinalityTagsDetector(MeterRegistry registry) {
+        this(registry, DEFAULT_THRESHOLD, DEFAULT_DELAY);
+    }
+
+    /**
+     * @param registry The registry to use to check the Meters in it
+     * @param threshold The threshold to use to detect high cardinality tags (if the
+     * number of Meters with the same name are higher than this value, that's a high
+     * cardinality tag)
+     * @param delay The delay between the termination of one check and the commencement of
+     * the next
+     */
+    public HighCardinalityTagsDetector(MeterRegistry registry, int threshold, Duration delay) {
+        this(registry, threshold, delay, DEFAULT_CONSUMER);
+    }
+
+    /**
+     * @param registry The registry to use to check the Meters in it
+     * @param threshold The threshold to use to detect high cardinality tags (if the
+     * number of Meters with the same name are higher than this value, that's a high
+     * cardinality tag)
+     * @param delay The delay between the termination of one check and the commencement of
+     * the next
+     * @param meterNameConsumer The action to execute if the first high cardinality tag is
+     * found
+     */
+    public HighCardinalityTagsDetector(MeterRegistry registry, int threshold, Duration delay,
+            Consumer<String> meterNameConsumer) {
+        this.registry = registry;
+        this.threshold = threshold;
+        this.delay = delay;
+        this.meterNameConsumer = meterNameConsumer;
+        this.scheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
```

Comment:
Would it be worth passing in a `io.micrometer.core.instrument.util.NamedThreadFactory` and give the poller thread a descriptive name?

Reply:
Definitely! Thanks!